cpp/src/parquet/arrow/schema.cc

```diff
@@ -689,10 +686,73 @@ Status GetOriginSchema(const std::shared_ptr<const KeyValueMetadata>& metadata,
 // but that is not necessarily present in the field reconstitued from Parquet data
 // (for example, Parquet timestamp types doesn't carry timezone information).
 
-Status ApplyOriginalStorageMetadata(const Field& origin_field, SchemaField* inferred) {
+Result<bool> ApplyOriginalMetadata(const Field& origin_field, SchemaField* inferred);
+
+using NestedFieldRewrapper = std::function<Status(FieldVector, SchemaField*)>;
+
+Status RewrapStructField(FieldVector new_children, SchemaField* inferred) {
+  inferred->field = inferred->field->WithType(::arrow::struct_(std::move(new_children)));
+  return Status::OK();
+}
+
+Status RewrapListField(FieldVector new_children, SchemaField* inferred) {
+  DCHECK_EQ(new_children.size(), 1);
+  inferred->field = inferred->field->WithType(::arrow::list(new_children[0]));
+  return Status::OK();
+}
+
+// XXX Perhaps add a DataType::WithFields method?
+NestedFieldRewrapper GetNestedFieldRewrapper(const ArrowType& origin_type,
+                                             const ArrowType& inferred_type) {
+  switch (inferred_type.id()) {
+    case ::arrow::Type::STRUCT:
+      if (origin_type.id() == ::arrow::Type::STRUCT) {
+        return RewrapStructField;
+      }
+      break;
+    case ::arrow::Type::LIST:
+      // TODO also allow LARGE_LIST and FIXED_SIZE_LIST
+      if (origin_type.id() == ::arrow::Type::LIST) {
+        return RewrapListField;
+      }
+      break;
+    default:
+      break;
+  }
+  return {};
+}
```

Comment:
IMHO it's more natural to provide a factory for data types:
```suggestion
std::function<std::shared_ptr<DataType>(FieldVector)> GetNestedFactory(
    const ArrowType& origin_type, const ArrowType& inferred_type) {
  switch (inferred_type.id()) {
    case ::arrow::Type::STRUCT:
      if (origin_type.id() == ::arrow::Type::STRUCT) {
        return ::arrow::struct_;
      }
      break;
    case ::arrow::Type::LIST:
      // TODO also allow LARGE_LIST and FIXED_SIZE_LIST
      if (origin_type.id() == ::arrow::Type::LIST) {
        return [](FieldVector fields) { return ::arrow::list(std::move(fields[0])); };
      }
      break;
    default:
      break;
  }
  return {};
}
```